File: 김소희/연산자 끼워넣기.py
N = int(input())
AList = list(map(int, input().split()))
opList = list(map(int, input().split()))

minCost = float('inf')
maxCost = -float('inf')

def dfs(index, cost):
    global minCost, maxCost
    if index == len(AList):
        minCost = min(minCost, cost)
        maxCost = max(maxCost, cost)
        return

    for i, op in enumerate(opList):
        if op > 0:
            opList[i] -= 1
            if i == 0:                
                dfs(index+1, cost + AList[index])
            elif i == 1:
                dfs(index+1, cost - AList[index])
            elif i == 2:
                dfs(index+1, cost * AList[index])
            elif i == 3:
                dfs(index+1, int(cost / AList[index]))
            opList[i] += 1

dfs(1, AList[0])
print(maxCost)
print(minCost)

File: 김소희/연산자 끼워넣기.java
import java.io.*;
import java.util.*;

public class Main {

    static int N;
    static int[] A;
    static int[] op;   // +, -, *, /
    static int minCost = Integer.MAX_VALUE;
    static int maxCost = Integer.MIN_VALUE;

    static void dfs(int index, int cost) {

        if (index == N) {
            minCost = Math.min(minCost, cost);
            maxCost = Math.max(maxCost, cost);
            return;
        }

        for (int i = 0; i < 4; i++) {
            if (op[i] > 0) {
                op[i]--;

                if (i == 0) {          // +
                    dfs(index + 1, cost + A[index]);
                } else if (i == 1) {   // -
                    dfs(index + 1, cost - A[index]);
                } else if (i == 2) {   // *
                    dfs(index + 1, cost * A[index]);
                } else if (i == 3) {   // /
                    dfs(index + 1, cost / A[index]); // Java는 0쪽으로 버림 → Python int()와 동일
                }

                op[i]++; // 복구
            }
        }
    }

    public static void main(String[] args) throws Exception {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        N = Integer.parseInt(br.readLine());

        A = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            A[i] = Integer.parseInt(st.nextToken());
        }

        op = new int[4];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 4; i++) {
            op[i] = Integer.parseInt(st.nextToken());
        }

        dfs(1, A[0]);

        System.out.println(maxCost);
        System.out.println(minCost);
    }
}
